File: app/services/tweet_share.py
from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from typing import Any, Literal

from app.formatters.telegram import TelegramPost, format_tweet
from app.providers.base import TweetData, TweetProvider, TweetProviderError
from app.utils.urls import ParsedTweetUrl, extract_first_tweet_url

ShareMode = Literal["private", "inline"]
ShareStatus = Literal["success", "error", "invalid_url"]

logger = logging.getLogger(__name__)


@dataclass(frozen=True, slots=True)
class ShareResult:
    status: ShareStatus
    tweet_id: str | None = None
    source_url: str | None = None
    normalized_url: str | None = None
    tweet: TweetData | None = None
    post: TelegramPost | None = None
    error_code: str | None = None
    elapsed_ms: int | None = None
    cache_hit: bool = False

    @property
    def ok(self) -> bool:
        return self.status == "success"


class TweetShareService:
    def __init__(
        self,
        *,
        provider: TweetProvider,
        cache_repository: Any,
        share_events_repository: Any,
        cache_ttl_seconds: int,
    ) -> None:
        self._provider = provider
        self._cache = cache_repository
        self._share_events = share_events_repository
        self._cache_ttl_seconds = cache_ttl_seconds

    async def process_text(
        self,
        text: str,
        *,
        telegram_user_id: int,
        chat_id: int | None,
        mode: ShareMode,
    ) -> ShareResult:
        parsed = extract_first_tweet_url(text)
        if not parsed:
            return ShareResult(status="invalid_url", error_code="invalid_url")
        return await self.process_url(
            parsed,
            telegram_user_id=telegram_user_id,
            chat_id=chat_id,
            mode=mode,
        )

    async def process_url(
        self,
        parsed: ParsedTweetUrl,
        *,
        telegram_user_id: int,
        chat_id: int | None,
        mode: ShareMode,
    ) -> ShareResult:
        started = time.perf_counter()
        cache_hit = False
        try:
            tweet = await self._cache.get(parsed.tweet_id)
            if tweet is None:
                tweet = await self._provider.get_tweet(parsed.tweet_id, parsed.normalized_url)
                await self._cache.set(
                    tweet,
                    parsed.source_url,
                    ttl_seconds=self._cache_ttl_seconds,
                )
            else:
                cache_hit = True

            post = format_tweet(tweet)
            elapsed_ms = _elapsed_ms(started)
            await self._share_events.create(
                telegram_user_id=telegram_user_id,
                chat_id=chat_id,
                tweet_id=parsed.tweet_id,
                source_url=parsed.source_url,
                mode=mode,
                status="success",
            )
            logger.info(
                "tweet_share status=success telegram_user_id=%s chat_id=%s "
                "tweet_id=%s mode=%s cache_hit=%s elapsed_ms=%s",
                telegram_user_id,
                chat_id,
                parsed.tweet_id,
                mode,
                cache_hit,
                elapsed_ms,
            )
            return ShareResult(
                status="success",
                tweet_id=parsed.tweet_id,
                source_url=parsed.source_url,
                normalized_url=parsed.normalized_url,
                tweet=tweet,
                post=post,
                elapsed_ms=elapsed_ms,
                cache_hit=cache_hit,
            )
        except TweetProviderError as exc:
            return await self._record_error(
                parsed,
                telegram_user_id=telegram_user_id,
                chat_id=chat_id,
                mode=mode,
                code=exc.code,
                started=started,
            )
        except Exception:
            logger.exception(
                "tweet_share status=error telegram_user_id=%s chat_id=%s "
                "tweet_id=%s mode=%s error_code=unexpected_error",
                telegram_user_id,
                chat_id,
                parsed.tweet_id,
                mode,
            )
            return await self._record_error(
                parsed,
                telegram_user_id=telegram_user_id,
                chat_id=chat_id,
                mode=mode,
                code="unexpected_error",
                started=started,
            )

    async def _record_error(
        self,
        parsed: ParsedTweetUrl,
        *,
        telegram_user_id: int,
        chat_id: int | None,
        mode: ShareMode,
        code: str,
        started: float,
    ) -> ShareResult:
        elapsed_ms = _elapsed_ms(started)
        await self._share_events.create(
            telegram_user_id=telegram_user_id,
            chat_id=chat_id,
            tweet_id=parsed.tweet_id,
            source_url=parsed.source_url,
            mode=mode,
            status="error",
            error_code=code,
        )
        logger.info(
            "tweet_share status=error telegram_user_id=%s chat_id=%s "
            "tweet_id=%s mode=%s error_code=%s elapsed_ms=%s",
            telegram_user_id,
            chat_id,
            parsed.tweet_id,
            mode,
            code,
            elapsed_ms,
        )
        return ShareResult(
            status="error",
            tweet_id=parsed.tweet_id,
            source_url=parsed.source_url,
            normalized_url=parsed.normalized_url,
            error_code=code,
            elapsed_ms=elapsed_ms,
        )


def _elapsed_ms(started: float) -> int:
    return round((time.perf_counter() - started) * 1000)

File: src/services/tweetShare.ts
import { formatThread, formatTweet, type TelegramPost } from "@/formatters/telegram";
import { TweetProviderError, type TweetData, type TweetProvider } from "@/providers/base";
import type { ShareEventRepository } from "@/repositories/shareEvents";
import type { TweetCacheRepository } from "@/repositories/tweetCache";
import { extractFirstTweetUrl, type ParsedTweetUrl } from "@/utils/urls";

export type ShareMode = "private" | "inline";
export type ShareStatus = "success" | "error" | "invalid_url";

// Errors that mean the tweet will never come back (deleted/private/not found):
// safe to negative-cache so we stop hammering providers. Transient errors
// (timeouts, rate limits, HTTP errors) are intentionally excluded.
const TERMINAL_ERROR_CODES = new Set(["not_found", "private_or_deleted"]);

export interface ShareResult {
  status: ShareStatus;
  ok: boolean;
  tweetId: string | null;
  sourceUrl: string | null;
  normalizedUrl: string | null;
  tweet: TweetData | null;
  post: TelegramPost | null;
  errorCode: string | null;
  elapsedMs: number | null;
  cacheHit: boolean;
  threadSize: number;
}

export interface ProcessOptions {
  telegramUserId: number;
  chatId: number | null;
  mode: ShareMode;
  /**
   * Force-disable thread unrolling for this request (e.g. the inline
   * "share single post" variant). Defaults to the service-wide setting.
   */
  unrollThread?: boolean;
}

export interface TweetShareService {
  processText(text: string, options: ProcessOptions): Promise<ShareResult>;
  processUrl(parsed: ParsedTweetUrl, options: ProcessOptions): Promise<ShareResult>;
}

interface Deps {
  provider: TweetProvider;
  cacheRepository: TweetCacheRepository;
  shareEventsRepository: Pick<ShareEventRepository, "create">;
  cacheTtlSeconds: number;
  negativeCacheTtlSeconds: number;
  threadUnrollEnabled: boolean;
  threadMaxTweets: number;
}

export function createTweetShareService(deps: Deps): TweetShareService {
  const {
    provider,
    cacheRepository,
    shareEventsRepository,
    cacheTtlSeconds,
    negativeCacheTtlSeconds,
    threadUnrollEnabled,
    threadMaxTweets,
  } = deps;

  // Fetch from the provider and write the result back to the cache (positive on
  // success, negative for terminal errors). Throws on any provider error.
  const fetchAndCache = async (
    tweetId: string,
    sourceUrl: string,
    normalizedUrl: string,
  ): Promise<TweetData> => {
    try {
      const tweet = await provider.getTweet(tweetId, normalizedUrl);
      await cacheRepository.set(tweet, sourceUrl, { ttlSeconds: cacheTtlSeconds });
      return tweet;
    } catch (error) {
      if (
        error instanceof TweetProviderError &&
        TERMINAL_ERROR_CODES.has(error.code) &&
        negativeCacheTtlSeconds > 0
      ) {
        await cacheRepository.setNegative(tweetId, sourceUrl, error.code, {
          ttlSeconds: negativeCacheTtlSeconds,
        });
      }
      throw error;
    }
  };

  // Load an ancestor while unrolling a thread. Any failure (missing, deleted,
  // provider error) stops the unroll rather than failing the whole share.
  const loadAncestor = async (tweetId: string): Promise<TweetData | null> => {
    const url = `https://x.com/i/status/${tweetId}`;
    try {
      const entry = await cacheRepository.getEntry(tweetId);
      if (entry?.kind === "hit") return entry.tweet;
      if (entry?.kind === "negative") return null;
      return await fetchAndCache(tweetId, url, url);
    } catch {
      return null;
    }
  };

  // Walk up the reply chain while the parent is the same author (a self-thread),
  // returning the tweets oldest → newest with the shared tweet last. X handles
  // are case-insensitive, and different providers/cached entries may disagree
  // on casing, so compare them case-insensitively.
  const sameAuthor = (a: TweetData, b: TweetData): boolean =>
    a.authorUsername.toLowerCase() === b.authorUsername.toLowerCase();

  const unrollThread = async (root: TweetData): Promise<TweetData[]> => {
    if (!threadUnrollEnabled || threadMaxTweets <= 1) return [root];
    let parent: TweetData | null = root.repliedToTweet;
    if (!parent || !sameAuthor(parent, root)) return [root];

    const ancestors: TweetData[] = [];
    while (parent && sameAuthor(parent, root)) {
      const current: TweetData = parent;
      ancestors.push(current);
      // Stop before fetching another ancestor once the cap (incl. root) is hit.
      if (ancestors.length + 1 >= threadMaxTweets) break;
      const parentId = current.inReplyToTweetId;
      if (!parentId) break;
      // The provider already fetched one level; reuse it instead of refetching.
      parent =
        current.repliedToTweet && current.repliedToTweet.tweetId === parentId
          ? current.repliedToTweet
          : await loadAncestor(parentId);
    }
    ancestors.reverse();
    return [...ancestors, root];
  };

  const recordError = async (
    parsed: ParsedTweetUrl,
    options: ProcessOptions,
    code: string,
    started: number,
    extra: { negativeCacheHit?: boolean } = {},
  ): Promise<ShareResult> => {
    const elapsedMs = elapsedSince(started);
    await shareEventsRepository.create({
      telegramUserId: options.telegramUserId,
      chatId: options.chatId,
      tweetId: parsed.tweetId,
      sourceUrl: parsed.sourceUrl,
      mode: options.mode,
      status: "error",
      errorCode: code,
    });
    log("error", {
      telegramUserId: options.telegramUserId,
      chatId: options.chatId,
      tweetId: parsed.tweetId,
      mode: options.mode,
      errorCode: code,
      negativeCacheHit: extra.negativeCacheHit ?? false,
      elapsedMs,
    });
    return {
      status: "error",
      ok: false,
      tweetId: parsed.tweetId,
      sourceUrl: parsed.sourceUrl,
      normalizedUrl: parsed.normalizedUrl,
      tweet: null,
      post: null,
      errorCode: code,
      elapsedMs,
      cacheHit: false,
      threadSize: 0,
    };
  };

  return {
    async processText(text, options): Promise<ShareResult> {
      const parsed = extractFirstTweetUrl(text);
      if (!parsed) {
        return invalidUrlResult();
      }
      return this.processUrl(parsed, options);
    },

    async processUrl(parsed, options): Promise<ShareResult> {
      const started = performance.now();
      let cacheHit = false;
      try {
        const entry = await cacheRepository.getEntry(parsed.tweetId);
        let tweet: TweetData;
        if (entry?.kind === "hit") {
          tweet = entry.tweet;
          cacheHit = true;
        } else if (entry?.kind === "negative") {
          // Known-bad tweet — short-circuit without touching the provider.
          return recordError(parsed, options, entry.errorCode, started, {
            negativeCacheHit: true,
          });
        } else {
          tweet = await fetchAndCache(parsed.tweetId, parsed.sourceUrl, parsed.normalizedUrl);
        }

        const wantThread = options.unrollThread !== false;
        const thread = wantThread ? await unrollThread(tweet) : [tweet];
        const post = thread.length > 1 ? formatThread(thread) : formatTweet(tweet);

        const elapsedMs = elapsedSince(started);
        await shareEventsRepository.create({
          telegramUserId: options.telegramUserId,
          chatId: options.chatId,
          tweetId: parsed.tweetId,
          sourceUrl: parsed.sourceUrl,
          mode: options.mode,
          status: "success",
        });
        log("success", {
          telegramUserId: options.telegramUserId,
          chatId: options.chatId,
          tweetId: parsed.tweetId,
          mode: options.mode,
          cacheHit,
          threadSize: thread.length,
          elapsedMs,
        });
        return {
          status: "success",
          ok: true,
          tweetId: parsed.tweetId,
          sourceUrl: parsed.sourceUrl,
          normalizedUrl: parsed.normalizedUrl,
          tweet,
          post,
          errorCode: null,
          elapsedMs,
          cacheHit,
          threadSize: thread.length,
        };
      } catch (error) {
        if (error instanceof TweetProviderError) {
          return recordError(parsed, options, error.code, started);
        }

        console.error("tweet_share unexpected error", error);
        return recordError(parsed, options, "unexpected_error", started);
      }
    },
  };
}

function invalidUrlResult(): ShareResult {
  return {
    status: "invalid_url",
    ok: false,
    tweetId: null,
    sourceUrl: null,
    normalizedUrl: null,
    tweet: null,
    post: null,
    errorCode: "invalid_url",
    elapsedMs: null,
    cacheHit: false,
    threadSize: 0,
  };
}

function elapsedSince(started: number): number {
  return Math.round(performance.now() - started);
}

function log(status: string, fields: Record<string, unknown>): void {
  console.log(`tweet_share status=${status}`, fields);
}
